### PR TITLE
Fix reverse_bits with registerless bits in all 3 circuit drawers

### LINKED ISSUE #7303: Circuit drawer reverse_bits does not reverse registerless bits
### Environment

- **Qiskit Terra version**: Current main
- **Python version**: 3.8
- **Operating system**: Ubuntu

### What is happening?

When using classical registers only with `reverse_bits`, the registers and the bits in the registers are reversed before display. However, bits without registers are reversed in terms of usage, but are not reversed when displayed.

```
bits = [Qubit(), Qubit(), Clbit(), Clbit()]
bits1 = [Clbit()]
cr = ClassicalRegister(2, "cr")
crx = ClassicalRegister(2, "crx")
qc = QuantumCircuit(bits, cr, [Clbit()], crx, bits1)
qc.x(0).c_if(crx[1], 0)
qc.measure(0, bits[3])
qc.draw('mpl', cregbundle=False, reverse_bits=True)
```
image

### How can we reproduce the issue?

---

### What should happen?

Registerless bits should show in descending order.
image

### Any suggestions?

This will require the merge of #7285 before the above circuit will work.

@javabster Can you assign this to me?

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

Fixes #7303

### Details and comments

This PR fixes an issue with the 3 circuit drawers where the registerless bits indexing was not reversed when the `reverse_bits` option was used. It now correctly displays as,
```
bits = [Qubit(), Qubit(), Clbit(), Clbit()]
bits1 = [Clbit()]
cr = ClassicalRegister(2, "cr")
cs = ClassicalRegister(2, "cs")
qc = QuantumCircuit(bits, cr, [Clbit()], cs, bits1)
qc.h(0).c_if(4, 0)
qc.measure(0, bits[3])
qc.draw('mpl', scale=0.7,cregbundle=False, reverse_bits=True)
```
image

### CODE REVIEW

File: qiskit/visualization/utils.py
Comment (jakelishman) [inline qiskit/visualization/utils.py:253]: Would this become any simpler by using the new `QuantumCircuit.find_bit`?  That handles bits in multiple registers and loose bits, and it's a constant-time lookup per bit (because `QuantumCircuit` stores the necessary data structure).  Either way, we might want to document somewhere appropriate what happens if a bit is in more than one register - the current behaviour looks like it'll get displayed as being in the last defined register.  I don't have a particular preference, so the current behaviour is fine, but we should document it.

For the future, likely we need to completely think how to display which bits are in which registers, but that's out of scope here.

**Review by jakelishman — COMMENTED** [review]
Thanks for doing this - the new form looks right.  The way that registers interact in the "order" with respect to loose bits is slightly unusual, but that's because of the super weird position `ClassicalRegister` is in right now in our data model - not the fault of this PR at all.

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py:234]: It should be easy to convert this to a `namedtuple`. I'll take a look at it.

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py:253]: Currently the 3 drawers do not have access to the `QuantumCircuit` object. `circuit_drawer` does have access, but it calls `get_layered_instructions` to create a `list(list(DAGOpNode))` which are the layers that will be displayed, and this is what is passed to the drawers. The `qregs, cregs, qubits, clbits` are passed as parameters to the drawers.

It would be easy enough to pass the `QuantumCircuit` object on down to the drawers. It also might make more sense to do the `_bit_locations` inside `get_layered_instructions` and pass the `BitLocations` object to the drawers at that time.

Never thought of the idea of a bit in more than one register. How does one decide??

Comment (jakelishman) [reply qiskit/visualization/utils.py:253]: Ah sure, that's no problem then.

Yeah, it's been a fairly big shift recently where `ClassicalRegister` has been completely demoted from its previous role of defining bits.  Now the individual bit objects (`Clbit`) are the primitive type, and `ClassicalRegister` (m.m. quantum) is just an alias to refer to several bits in one go.  It's not something we need to worry about in this PR, but at some point we might want to think about how best to convey that information visually.

**Comment by enavarro51** [comment]
@javabster @jakelishman Based on the comment above about using `QuantumCircuit.find_bit` this PR was updated to make the drawers much more 'bit forward'. So the latest changes convert from passing `qregs` and `cregs` to the drawers and use only `qubits` and `clbits`. The `circuit` instance is passed  and registers are then accessed through `find_bit` which also returns the `index` in the circuit. This made it possible to move much of the code for getting register/index to `utils.py`.

This required updating code sections for displaying register and bit labels and for measures and conditions. There are still some unresolved areas which should go into a new issue. These include what should the label look like for a registerless bit, how to handle bits that belong to multiple registers (currently only the first register in the list is used), and how to display conditions on a list of unrelated bits.

**Comment by jakelishman** [comment]
Sorry I've not looked at this properly - the new form certainly looks a lot cleaner, but I'll have to have a proper look through in the new year.  Do you have any ideas for how to display overlapping registers in the drawers?  I wonder if we could do something like coloured square brackets on the side (if we need to) - we could stick out arbitrarily far on the left to do that, perhaps?

**Comment by enavarro51** [comment]
@jakelishman I have a follow-on to this ready which will take the changes in #7423 and convert all the drawers to using `QuantumCircuit.reverse_bits` instead of doing their own reverse process. At that point the outputs of `qc.reverse_bits` and `qc.draw(reverse_bits=True)` will always be the same. If you do both, you get the original circuit drawn.

I built the new drawer `reverse_bits` on top of this PR, so as soon as this and #7423 are merged, I can submit the new PR. Thanks.

File: qiskit/visualization/matplotlib.py
Comment (javabster) [inline qiskit/visualization/matplotlib.py:168]: unclear why `_q_anchors` and `_c_anchors` are being removed here? I see they haven't been removed elsewhere, just here

File: qiskit/visualization/text.py
Comment (javabster) [inline qiskit/visualization/text.py:778]: Is this comment still relevant?

File: qiskit/visualization/utils.py
Comment (javabster) [inline qiskit/visualization/utils.py:166]:
```suggestion
    The key to the dict is either the (Qubit, Clbit) or if cregbundle True,
```

File: qiskit/visualization/utils.py
Comment (javabster) [inline qiskit/visualization/utils.py:434]: Another small thing, not essential now but for next time, I think it's generally better for new args and return values to be added at the end of the line to prevent unintentional errors where it's used elsewhere

**Review by javabster — COMMENTED** [review]
As I've been away I haven't been following the progress on this closely, and there is a lot of scope creep, but I don't see any major issues. Have left a few minor comments but Jake may have more.

Also when I ran the snapshot tests there were failures for `test_iqx_colors` and `test_plot_bloch_multivector` that need resolving.

Thanks for your contributions Ed! 😄

**Comment by enavarro51** [comment]
> Also when I ran the snapshot tests there were failures for `test_iqx_colors` and `test_plot_bloch_multivector` that need resolving.

@javabster These 2 fail on main as well. I can update the references, but I don't know where the failures crept in.

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:89]: Returned in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/55e62189a965a13f2fee8bdb9cdfb1f9cb88d3fd.

File: qiskit/visualization/matplotlib.py
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:168]: These were just moved up above at line 113.

File: qiskit/visualization/text.py
Comment (enavarro51) [reply qiskit/visualization/text.py]: Removed in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/55e62189a965a13f2fee8bdb9cdfb1f9cb88d3fd.

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py]: Fixed in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/55e62189a965a13f2fee8bdb9cdfb1f9cb88d3fd.

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py:434]: Got it.

**Comment by enavarro51** [comment]
Looks like the last step in the iqx-dark PR was to change the plus in the cnot from white to background, but the ref image was not updated, There seems to be a color diff for the arrow in the bloch one. but not sure why the color is different. Both image refs updated in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/55e62189a965a13f2fee8bdb9cdfb1f9cb88d3fd.

File: qiskit/visualization/matplotlib.py
Comment (javabster) [reply qiskit/visualization/matplotlib.py:168]: ah got it 😅

**Review by javabster — DISMISSED** [review]
Thanks for the changes Ed, when Jake get's round to re-reviewing he might have some more comments but this LGTM. I don't see any issues when testing manually and the snapshot tests all pass 🚀

File: qiskit/visualization/matplotlib.py
Comment (jakelishman) [inline qiskit/visualization/matplotlib.py:275]: If we're changing the name of this function, can we make it a bit clearer that it also sets a bunch of properties on `self`?  I was confused why there was so much logic in a function that seemingly just returns the number of lines of something.

(`_set_bit_reg_info` would be better than `_get`, but there may be a better name as well)

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:58]: Ok. First 2 seem straightforward. Will have to look at the 3rd.

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py:183]: Will do.

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py:231]: Got it.

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py:321]: Cool. That looks simpler and more readable.

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:194]: Right. The key to the `_bits_regs_map` will be a Qubit, Clbit, or ClassicalRegister, the last only if cregbundle is on. This way the indexes match the indexes of the displayed wires. Can change it to `wires` where appropriate.

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:202]: Good point. Should be able to change that universally.

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:311]: Ok.

File: qiskit/visualization/matplotlib.py
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:275]: I like `set` in this instance. It's one of the rare functions what just sets up 2 dicts, which are then used later to do the actual drawing.

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:125]: This is from `circuit_visualization.py`. I'm not sure what moving the setting of `layout` and `global_phase` into the drawers would buy you.
```
    if with_layout:
        layout = circuit._layout
    else:
        layout = None

    if fold is None:
        fold = 25

    global_phase = circuit.global_phase if hasattr(circuit, "global_phase") else None
    qcd = _matplotlib.MatplotlibDrawer(
        qubits,
        clbits,
        nodes,
        scale=scale,
        style=style,
        reverse_bits=reverse_bits,
        plot_barriers=plot_barriers,
        layout=layout,
        fold=fold,
        ax=ax,
        initial_state=initial_state,
        cregbundle=cregbundle,
        global_phase=global_phase,
        calibrations=circuit.calibrations,
        circuit=circuit,
    )
```

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:603]: This section only occurs when the `condition `is a whole register, and the `cond_len` is the size of the register minus 1. `cwire` and `gap` are guaranteed to be larger than `cond_len`, so `\cwx[--2]` won't occur.

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:58]: How about this as a solution. One thing to keep in mind is that the drawer classes will not work unless `_get_layered_instructions` has been called previously and the returns from that passed to the drawer.
- First off `qregs `and `cregs `are superfluous in the old and the new code. The old code still uses them, but the new code shows that everything can be done with `qubits `and `clbits`.
- Leave `qregs `and `cregs `in the drawer classes.
- Pass `qregs=None` and `cregs=None` from all the calls to the drawers from `circuit_visualization.py`.
- If a drawer detects a not None `qregs `or `cregs`, do a standard deprecation warning. The values in them will be ignored. `qregs`, `cregs` and `qubits`, `clbits `must be in sync or nothing works.
- If a drawer detects `circuit == None`, then create a new empty `QuantumCircuit` instance using `qubits `and `clbits`, and warn the user that the circuit needs to be passed in. The circuit is only used for the `find_bits` currently, so the `_data` portion is ignored.

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/65d90e68c87a048ad56dde08f2aa2a74855e215f

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/65d90e68c87a048ad56dde08f2aa2a74855e215f.

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/65d90e68c87a048ad56dde08f2aa2a74855e215f

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/65d90e68c87a048ad56dde08f2aa2a74855e215f

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/65d90e68c87a048ad56dde08f2aa2a74855e215f

File: qiskit/visualization/matplotlib.py
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/7319/commits/65d90e68c87a048ad56dde08f2aa2a74855e215f

File: qiskit/visualization/utils.py
Comment (jakelishman) [reply qiskit/visualization/utils.py]: As it turns out, I was mistaken that `register.index` was equivalent to `list.index` (`register` specialises it into a dict lookup as opposed to the linear-time search), but the new way is a slightly cleaner anyway.  Sorry about that - should have checked my nitpicks first.

File: qiskit/visualization/latex.py
Comment (jakelishman) [reply qiskit/visualization/latex.py:603]: Ok, that makes sense.

File: qiskit/visualization/utils.py
Comment (enavarro51) [reply qiskit/visualization/utils.py:327]: I believe this was handled in #7459.

File: qiskit/visualization/latex.py
Comment (jakelishman) [reply qiskit/visualization/latex.py:58]: Yeah, that sounds like it matches what I had in mind as well.  I think there are a few other parts of the constructor that could similarly get this treatment now - `global_phase` seems like an obvious candidate - and things like `qubits` and `clbits` should eventually not be necessary, since they can also be read from the circuit.

For now, let's just do what you suggested with `qregs`, `cregs` and `global_phase`.

File: qiskit/visualization/latex.py
Comment (jakelishman) [reply qiskit/visualization/latex.py:125]: It just buys you better data consistency; there's a single source of truth, and no opportunity for them to get out of sync if the code is expanded.  The data's already there in the class, so if we were re-writing the function, the question would be "why do we separate them out in the call?" - it's just that the development didn't quite go that way round.

File: qiskit/visualization/utils.py
Comment (jakelishman) [reply qiskit/visualization/utils.py:327]: I don't think so - I think you need to do `label = get_bit_label(...) + f"={cond_val:#x}" or something to get the escaping after that?

**Review by jakelishman — COMMENTED** [review]
Aside from the couple of outstanding comments that had previously been waiting for me to give the go-ahead, this all looks sensible to me, thanks.

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:58]: Ok. I'll get the deprecations set up for these 3.

File: qiskit/visualization/latex.py
Comment (enavarro51) [reply qiskit/visualization/latex.py:125]: Ok. How about this. We do a new issue to push setting `qubits, clbits,, and layout` into the drawers along with appropriate deprecations, using `circuit` that's being passed into the drawers.

Comment (enavarro51) [reply qiskit/visualization/latex.py:125]: Actually looks like `calibrations` can be pushed as well.

Comment (enavarro51) [reply qiskit/visualization/latex.py:125]: So I think it makes sense to do everything except `qubits `and `clbits` here, since the deprecation process will all be similar. We'll put off `qubits` and `clbits ` until later since they might be a bit more complicated because of `idle_wires=False`. That's `cregs`, `qregs`, `global_phase`, `layout`, and `calibrations`.

Comment (enavarro51) [reply qiskit/visualization/latex.py:125]: Ok. One more issue. If we deprecate `layout` as an arg to the drawer classes, we'd need to pass the `with_layout` bool to determine whether to use the `layout`. So which is preferred?
```
    if with_layout:
        layout = circuit._layout
    else:
        layout = None
```
from `circuit_visualization.py`.

Comment (jakelishman) [reply qiskit/visualization/latex.py]: Your plan sounds good to me. I prefer passing `with_layout`, because it feels like a drawing option, whereas since `QuantumCircuit._layout` is a private field, it's better for users not to need to access it.

**Comment by enavarro51** [comment]
@jakelishman I have now deprecated `qregs, cregs, global_phase, layout, and calibrations` from the 3 circuit drawers and these are now set using the newly passed in `circuit` kwarg. If `circuit` is `None`, a new circuit is created in the `__init__` of the drawer using the `qubits` and `clbits` args. I also added `with_layout` as a kwarg and this is used to decide whether to use the `QuantumCircuit._layout` or not. I believe this covers all the outstanding review issues.

File: qiskit/visualization/latex.py
Comment (1ucian0) [inline qiskit/visualization/latex.py:64]: In general, it is better to add parameters at the end, in case somebody is calling the method with positional arguments. I dont think is a biggy in this case. 

```suggestion
        initial_state=False,
        cregbundle=False,
        global_phase=None,
        qregs=None,
        cregs=None,
        with_layout=False,
        circuit=None,
```

File: qiskit/visualization/latex.py
Comment (1ucian0) [inline qiskit/visualization/latex.py:126]: Is this a *deprecation* warning? What is being replaced with in the future?

**Review by 1ucian0 — APPROVED** [review]
Thanks Edwin!

**Pull request merged**
